sw/device/lib/drivers/gpio/gpio.c

```diff
@@ -1,213 +1,525 @@
+ */
+#define GPIO_CFG_INTR_MODE_INDEX    0
+
+
+/****************************************************************************/
+/**                                                                        **/
+/*                        TYPEDEFS AND STRUCTURES                           */
+/**                                                                        **/
+/****************************************************************************/
+
+/****************************************************************************/
+/**                                                                        **/
+/*                      PROTOTYPES OF LOCAL FUNCTIONS                       */
+/**                                                                        **/
+/****************************************************************************/
+
+/****************************************************************************/
+/**                                                                        **/
+/*                           EXPORTED VARIABLES                             */
+/**                                                                        **/
+/****************************************************************************/
+
+/****************************************************************************/
+/**                                                                        **/
+/*                            GLOBAL VARIABLES                              */
+/**                                                                        **/
+/****************************************************************************/
+
+/****************************************************************************/
+/**                                                                        **/
+/*                           EXPORTED FUNCTIONS                             */
+/**                                                                        **/
+/****************************************************************************/
+
+gpio_result_t gpio_config (gpio_cfg_t cfg){
+    /* check that pin is in acceptable range. */
+    if (cfg.pin > (MAX_PIN-1) || cfg.pin < 0)
+        return GpioPinNotAcceptable;
+    /* reset pin coniguration first.*/
+    gpio_reset (cfg.pin);
+    /* check mode. */
+    if ((cfg.mode < GpioModeIn) || (cfg.mode > GpioModeoutOpenDrain1))
+        return GpioModeNotAcceptable;
+    /* set mode. */
+    gpio_set_mode (cfg.pin, cfg.mode);
+    /* if input sampling is enabled */
+    if (cfg.en_input_sampling == true)
+        gpio_en_input_sampling (cfg.pin);
+    /* if interrupt is enabled. Also after enabling check for error */
+    if (cfg.en_intr == true)
+        if (gpio_intr_en (cfg.pin, cfg.intr_type) != GpioOk)
+            return GpioIntrTypeNotAcceptable;
+    return GpioOk;
+}
+
+gpio_result_t gpio_set_mode (gpio_pin_number_t pin, gpio_mode_t mode)
+{
+    if (pin >= 0 && pin < 16)
```

Comment:
can we add here a define based on the max number of pins? (imagine that in the future they change...)

Reply:
I added it in x-heep.h, however, im not checking at the moment for more than 32b. Should I extend all the functionalities for the future with checking the define in the functions?!

Reply:
let's solve this question tomorrow